### PR TITLE
API compatibility

### PR DESCRIPTION
**Pull request by wence-**
Some more API compatibility.  Mostly to support new firedrake-adjoint.

### CODE REVIEW

File: firedrake/bcs.py [hunk 1]
```diff
@@ -159,6 +160,13 @@ def set_value(self, val):
         self._original_val = val
 
     @utils.cached_property
+    def domain_args(self):
+        """The sub_domain the BC applies to."""
+        if isinstance(self.sub_domain, str):
+            return (self.sub_domain, )
+        return (as_tuple(self.sub_domain), )
```
Comment (wence-) [inline firedrake/bcs.py]: Arguably this is wrong, but it works for the use-cases we need.

Additionally, we should also provide a copy constructor for DirichletBC.

File: firedrake/constant.py [hunk 2]
```diff
@@ -87,6 +87,10 @@ def evaluate(self, x, mapping, component, index_values):
             return self.dat.data_ro
         return self.dat.data_ro[component]
 
+    def values(self):
+        """Return a (flat) view of the value of the Constant."""
+        return self.dat.data_ro.reshape(-1)
+
     def function_space(self):
         """Return a null function space."""
         return None
```
Comment (wence-) [inline firedrake/constant.py]: Arguably, we should go back and make `Constant(...) => Function(FunctionSpace(mesh, "Real", 0))`.  But the implementation of the latter does not support vector and tensor-valued things, whereas the former does.

File: firedrake/function.py [hunk 3]
```diff
@@ -232,7 +234,8 @@ def __init__(self, function_space, val=None, name=None, dtype=ScalarType):
             raise NotImplementedError("Can't make a Function defined on a "
                                       + str(type(function_space)))
 
-        if isinstance(val, CoordinatelessFunction):
+        if isinstance(val, (Function, CoordinatelessFunction)):
+            val = val.topological
             if val.function_space() != V.topological:
                 raise ValueError("Function values have wrong function space.")
             self._data = val
```
Comment (wence-) [inline firedrake/function.py]: Allow creating a new function sharing data with an existing one.

File: firedrake/functionspaceimpl.py [hunk 4]
```diff
@@ -391,19 +391,25 @@ def __mul__(self, other):
...
         """The number of degrees of freedom (includes halo dofs) of this
         function space on this process. Cf. :attr:`node_count`."""
-        return self.node_count*self.dim
+        return self.node_count*self.shape_dim
+
+    def dim(self):
+        """The global number of degrees of freedom for this function space.
+
+        See also :attr:`dof_count` and :attr:`node_count`."""
+        return self.dof_dset.layout_vec.getSize()
```
Comment (wence-) [inline firedrake/functionspaceimpl.py]: Fixing #884.

File: firedrake/ufl_expr.py [hunk 5]
```diff
@@ -138,15 +138,24 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
 
     See also :func:`ufl.derivative`.
     """
-    if set(extract_coefficients(form)) & set(u.split()):
+    # TODO: What about Constant?
+    if len(u.split()) > 1 and set(extract_coefficients(form)) & set(u.split()):
```
Comment (wence-) [inline firedrake/ufl_expr.py]: This check is ugly.  If `f` is not a mixed function, then `f.split()` is a singleton tuple, which contains a new function that shares data with `f`.  Whereas if `f` is a `Constant`, then `f.split()` is a singleton tuple which is `f`.

File: firedrake/ufl_expr.py [hunk 6]
```diff
@@ -138,15 +138,24 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
...
             number = max(a.number() for a in args) if args else -1
             du = Argument(V, number + 1)
+        elif isinstance(u, firedrake.Constant):
+            if u.ufl_shape != ():
+                raise ValueError("Real function space of vector elements not supported")
+            mesh = form.ufl_domain()
+            V = firedrake.FunctionSpace(mesh, "Real", 0)
+            args = form.arguments()
+            number = max(a.number() for a in args) if args else -1
+            du = Argument(V, number + 1)
```
Comment (wence-) [inline firedrake/ufl_expr.py]: Should be able to differentiate wrt `Constant`.  Again, suggests `Constant` should just be `Function(FunctionSpace(mesh, "Real", 0))`.

File: firedrake/extrusion_utils.py [hunk 7]
```diff
@@ -58,7 +58,7 @@ def make_extruded_coords(extruded_topology, base_coords, ext_coords,
                 ext_coords[2*d+1][%(base_coord_dim)d] = *layer_height * (layer[0][0] + 1);
             }
         }""" % {'base_map_arity': base_coords.cell_node_map().arity,
-                'base_coord_dim': base_coords.function_space().dim},
+                'base_coord_dim': base_coords.function_space().shape_dim},
```
Comment (miklos1) [inline firedrake/extrusion_utils.py]: `value_size`?

File: firedrake/functionspaceimpl.py (hunk 4 above)
Comment (miklos1) [inline firedrake/functionspaceimpl.py]: :+1:

File: firedrake/extrusion_utils.py (hunk 7 above)
Comment (wence-) [inline firedrake/extrusion_utils.py]: `.ufl_element().value_size()`.  Yes, that would work, but breaks if coordinates become intrinsically vector-valued (unlikely).

Comment (miklos1) [inline firedrake/extrusion_utils.py]: `.ufl_element().value_size()` is not quite the same as you mentioned. What I wanted to suggest is that `value_size` might be a more consistent/familiar name than `shape_dim`.

Comment (wence-) [inline firedrake/extrusion_utils.py]: Oh, I see, yes.

**Pull request merged**